src/main/java/com/android/volley/CacheDispatcher.java

```diff
@@ -154,4 +179,92 @@ public void run() {
             }
         }
     }
+
+    /** Request received a valid response that can be used by other waiting requests. */
+    @Override
+    public void onResponseReceived(Request<?> request, Response<?> response) {
+        if (response.cacheEntry == null || response.cacheEntry.isExpired()) {
+            onNoUsableResponseReceived(request);
+            return;
+        }
+        String cacheKey = request.getCacheKey();
+        Queue<Request<?>> waitingRequests;
+        synchronized (mWaitingRequests) {
+            waitingRequests = mWaitingRequests.remove(cacheKey);
+        }
+        if (waitingRequests != null) {
+            if (VolleyLog.DEBUG) {
+                VolleyLog.v("Releasing %d waiting requests for cacheKey=%s.",
+                    waitingRequests.size(), cacheKey);
+            }
+            // Process all queued up requests.
+            for (Request<?> waiting : waitingRequests) {
+                mDelivery.postResponse(waiting, response);
+            }
+        }
+    }
+
+    /** No valid response received from network, release waiting requests. */
+    @Override
+    public void onNoUsableResponseReceived(Request<?> request) {
+        String cacheKey = request.getCacheKey();
+        Queue<Request<?>> waitingRequests;
+        synchronized (mWaitingRequests) {
+            waitingRequests = mWaitingRequests.remove(cacheKey);
+        }
+        if (waitingRequests != null) {
```

Comment:
This is racy and you should expand the synchronized block above to also include the mWaitingRequests.put() call below. If you don't then the map can be modified by maybeAddToWaitingRequests() before the code hits the synchronized block below and any changes will be clobbered.

Reply:
Done.